delta/sourcing-psql/src/main/scala/ch/epfl/bluebrain/nexus/delta/sourcing/stream/Pipe.scala

```diff
@@ -0,0 +1,166 @@
+package ch.epfl.bluebrain.nexus.delta.sourcing.stream
+
+import ch.epfl.bluebrain.nexus.delta.rdf.IriOrBNode.Iri
+import ch.epfl.bluebrain.nexus.delta.sourcing.model.{Envelope, Label}
+import ch.epfl.bluebrain.nexus.delta.sourcing.stream.Elem.{DroppedElem, FailedElem, SuccessElem}
+import ch.epfl.bluebrain.nexus.delta.sourcing.stream.ElemCtx.SourceIdPipeChainId
+import ch.epfl.bluebrain.nexus.delta.sourcing.stream.ProjectionErr.PipeInOutMatchErr
+import fs2.Pull
+import monix.bio.Task
+import shapeless.Typeable
+
+/**
+  * Pipes represent individual steps in a [[Projection]] where [[SuccessElem]] values are processed to produce other
+  * [[Elem]] values. They apply to [[Envelope]] of [[SuccessElem]] of type [[Pipe#In]] and produce [[Envelope]] of
+  * [[Elem]] of type [[Pipe#Out]].
+  *
+  * Pipes can perform all sorts of objectives like filtering and transformations of any kind.
+  *
+  * They are ultimately chained and attached to sources to complete the underlying projection [[Stream]] that is run.
+  */
+trait Pipe { self =>
+
+  /**
+    * The underlying element type accepted by the Pipe.
+    */
+  type In
+
+  /**
+    * The underlying element type emitted by the Pipe.
+    */
+  type Out
+
+  /**
+    * @return
+    *   the label that represents the specific pipe type
+    */
+  def label: Label
+
+  /**
+    * @return
+    *   the name of the pipe, coinciding with its label before joining with other sources and pipes
+    */
+  def name: String = label.value
+
+  /**
+    * @return
+    *   the Typeable instance for the accepted element type
+    */
+  def inType: Typeable[In]
+
+  /**
+    * @return
+    *   the Typeable instance for the emitted element type
+    */
+  def outType: Typeable[Out]
+
+  /**
+    * The pipe behaviour left abstract to be implemented by each specific pipe. It takes an element of type In that up
+    * to this pipe has been processed successfully and returns another element (possibly failed, dropped) of type Out.
+    * @param element
+    *   the element to process
+    * @return
+    *   a new element (possibly failed, dropped) of type Out
+    */
+  def apply(element: Envelope[Iri, SuccessElem[In]]): Task[Envelope[Iri, Elem[Out]]]
```

Comment:
We will see how things go and it can be done later but I would advocate to drop Envelope there and have the relevant things directly in `Elem` or its context